pkg/virt-handler/vm.go

```diff
@@ -324,7 +343,7 @@ func (d *VirtualMachineController) execute(key string) error {
 
 	// Determine removal of VirtualMachineInstance from cache should result in deletion.
 	if !vmiExists {
-		if domainExists {
+		if domainAlive {
```

Comment:
We still need to delete the domain even if it isn't alive. This is why we're getting stale domains in that if statement above that don't match the current VMI.

Reply:
That's what I don't want anymore, I want to keep the domain in order to set the correct final status to the VMI. I tried to postpone the domain deletion to after setting the state, but then virt-launcher is already down (and so the domain is gone anyway)

Reply:
I see now. I think the problem we're having is that the call to ```client.KillVirtualMachine(vmi)``` has been involved with both killing the libvirt domain and Undefining it. We've lost the Undefine() part in this new logic because we no longer call ```client.KillVirtualMachine(vmi)``` on domains that are not alive. 

The domain being Undefined() acts as an acknowledgement to virt-launcher that virt-handler is done with the domain. The result is virt-launcher then sends a delete notification (which will remove the domain from the virt-handler cache) and then virt-launcher know's it is safe to exit.

The fix here i that we need the shutdown flow breakup the domain shutdown from the final Undefine step. Once virt-handler sees that the domain is down and that the VMI has the correct status reported, then there should be a final call to Undefine the VM which gracefully free's virt-launcher to exit.

Reply:
That fix sounds like a good idea, thanks. Will give it a try tomorrow.